### ISSUE TITLE
[swift] Error when calling `getAccessToken` on a FirefoxAccount restored from JSON

### ISSUE THREAD

**Issue by sashei** [mozilla/application-services#1195]
We are currently working to get more detailed logs for y'all around this issue. Filing for now to capture basic incident.

I'm seeing two errors coming through from this particular ticket:
- `FxAException: Unauthorized error: No cached token for scope https://identity.mozilla.com/apps/oldsync` <-- more frequent
- `FxAException: Unauthorized error: Remote server error: '401' '110' 'Unauthorized' 'Invalid authentication token in request si...` (this got cut off in our error reporter)

┆Issue is synchronized with this [Jira Story](https://jira.allizom.org/browse/SYNC-32)
┆Epic: Lockwise Support

**Comment by vladikoff** [comment]
Thanks for filing this issue!

**Comment by jdragojevic** [comment]
Is fixing this considered urgent?

**Comment by linuxwolf** [comment]
TL;DR - We think it's important, but not urgent.

We've worked around the fatal crashes on our side, so we're not blocked on updates to fix it.  However, part of handling it results in resetting the user's state (forcing a new sign-in).  Understanding what's going on is important to give the user a better error message, if we can't avoid the situation on our side.

**Comment by sashei** [comment]
@vladikoff I've included some more detailed error messages in the ticket description.

**Comment by rfk** [comment]
> Unauthorized error: Remote server error: '401' '110' 'Unauthorized' 'Invalid

So this could be caused by e.g. the user going into their "devices and apps" list and revoking the Lockbox token, or by the user doing a password reset.  I'd expect these to throw a `FirefoxAccountError.Unauthorized` per the comment here:

  https://github.com/mozilla/application-services/blob/8432a5a0e4a627cf0ceb7eea72a23168e99a96ac/components/fxa-client/ios/FxAClient/FirefoxAccount.swift#L238

Which the app should catch and either enter some "please reconnect to your account" state (like Desktop Firefox does) or behave as though the user signed out (which IIUC is what Fennec currently does).

**Comment by rfk** [comment]
From meeting: it feels like the "Unauthorized error: No cached token for scope https://identity.mozilla.com/apps/oldsync" error is happening more often than could be explain by password resets or other disconnect events.  IIUC there are two cases where we might produce this error message:

* When you have a refresh token, but it doesn't have the right scopes (https://github.com/mozilla/application-services/blob/master/components/fxa-client/src/oauth.rs#L49)
* When you don't have a refresh token at all (https://github.com/mozilla/application-services/blob/master/components/fxa-client/src/oauth.rs#L66)

I have a hard time imagining a scenario that leads to the first, because we don't provide a way to edit the scopes associated with a refresh token. So I'm guessing it's more to do with the second.

One thing we could do here is provide more granular message messages for these cases (neither of which are accurately described by "no cached access token").

**Comment by rfk** [comment]
> When you don't have a refresh token at all 

One reason this could happen would be if we don't correctly migrate from the V1 to the V2 account state format, via https://github.com/mozilla/application-services/blob/master/components/fxa-client/src/state_persistence.rs#L59.

If so, I would expect to see the rate of these errors decline over time, as users are forced to re-login and save their data in the new format. Can we check whether the errors are becoming less frequent over time?

**Comment by cse-test-sync** [comment]
➤ Vladislav Filippov commented:

from mtg: need more info

**Comment by vladikoff** [comment]
Please reopen if we have more information
